app/controllers/application_controller.rb

```diff
@@ -5,6 +5,176 @@ class ApplicationController < Sinatra::Base
+          redirect '/tweets'
+      else 
+      erb :'users/create_user'
+    end 
+  end
+
+  post '/signup' do 
+   
+      if params[:username].present? && params[:email].present? && params[:password].present?
+        @user = User.create(username: params[:username], email: params[:email], password: params[:password])
+        session[:user_id] = @user.id
+      
+        redirect '/tweets'
+      else 
+        redirect '/signup'
+      
+    end 
+  end 
+
+  get '/tweets' do 
+ 
+    if is_logged_in?
+        @fweets = Tweet.all
+        @posts = []
+        @fweets.each do |f|
+        @posts << {username: User.find(f.user_id).username, content: f.content, user_id: f.user_id, id: f.id}
+          end
+      @user = User.find(session[:user_id])
+      erb :'tweets/tweets' 
+    else
+     redirect '/login'
+    end 
+  end 
+
+  get '/tweets/new' do 
+      if is_logged_in?
+        erb :'/tweets/create_tweet'
+      else 
+        redirect "/login"
+    end
+  end 
+
+  post '/tweets/new' do 
+   if !params[:tweet].present?
+    redirect '/tweets/new'
+  end 
+
+    if is_logged_in? && params[:tweet].present?
+      @user = User.find(session[:user_id])
+      @tweet = Tweet.create(content: params[:tweet], user_id: @user.id)
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end 
+  end 
+
+
+  post '/login' do
+    @user = User.find_by(:username => params[:username])
+        if @user && @user.authenticate(params[:password])
+              session[:user_id] = @user.id
+        redirect '/tweets'
+      else 
+      erb :'/users/login'
+    end 
+  end 
+
+  get '/tweets/:id' do 
+   
+      if is_logged_in?
+        @tweet = Tweet.find(params[:id])
+     
+        erb :'/tweets/one_tweet'
+      else 
+        redirect '/login'
+      end
+    end
+
+  get '/logout' do 
+    session.clear 
+    redirect '/login'
+  end 
+
+  get '/users/:slug' do 
+    
+    @user = User.find_by(username: params[:slug])
+      @fweets = Tweet.where(user_id: @user.id)
+      erb :'/tweets/show_tweet'
+  end 
+
+
+  get '/tweets/:id/edit' do 
+    binding.pry
+    if is_logged_in? && User.find(Tweet.find(params[:id]).user_id) == User.find(session[:user_id])
+    @tweet = Tweet.find(params[:id])
+    erb :'tweets/edit_tweet'
+  else 
+      redirect '/login'
+    end 
+  end 
+
+  post '/tweets/:id/edit' do 
+  
+    
+  if params[:tweet].present?
+    @tweet= Tweet.find(params[:id])
+    @tweet.update(content: params[:tweet])
+
+    redirect '/tweets'
+  else 
+    redirect "/tweets/#{@tweet.id}/edit"
+  end 
+  end
+
+
+  get '/tweet/:id/delete' do 
+    
+  if is_logged_in? && User.find(Tweet.find(params[:id]).user_id) == User.find(session[:user_id])
+    Tweet.find(params[:id]).delete
+    redirect '/tweets'
+  else 
+    redirect '/tweets'
+  end 
+  end 
+
+  post '/tweet/:id/delete' do 
+  
+    if is_logged_in? && User.find(Tweet.find(params[:id]).user_id) == User.find(session[:user_id])
+      @tweet = Tweet.find(params[:id])
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.